File: Exercism/javascript/pythagorean-triplet/pythagorean-triplet.js
'use strict'

const sum = (a, b) => a + b
const product = (a, b) => a * b

const numbers = function * (from, too = Number.POSITIVE_INFINITY, step = 1) {
  for (let i = from; i < too; i += step) {
    yield i
  }
}

const GENTRIPLES = function * (max) {
  for (const a of numbers(1, max)) {
    for (const b of numbers(a, max)) {
      const t = new Triplet(a, b, Math.hypot(a, b))
      if (t.isPythagorean()) {
        yield t
      }
    }
  }
}

const filter = function * (iterable, ...fns) {
  for (const value of iterable) {
    if (fns.every((fn) => fn(value))) {
      yield value
    }
  }
}

export class Triplet {
  constructor (...triple) {
    this.triple = triple.sort((a, b) => a - b)
  }

  sum () {
    return this.triple.reduce(sum)
  }

  product () {
    return this.triple.reduce(product)
  }

  isPythagorean () {
    const t = this.triple
    const [a, b, c] = t
    return t.length === 3 &&
      t.every(Number.isInteger) &&
      a ** 2 + b ** 2 === c ** 2
  }

  static where ({ sum, minFactor = 0, maxFactor }) {
    if (!maxFactor) {
      throw new Error('maxFactor not set')
    }
    const triplets = filter(
      GENTRIPLES(maxFactor),
      (t) => t.triple.every((v) => v > minFactor)
    )
    return sum && sum > 0
      ? [...filter(triplets, (t) => t.sum() === sum)]
      : [...triplets]
  }
}

File: Exercism/python/pythagorean-triplet/pythagorean_triplet.py
def triplets_with_sum(number):
    return {triple for triple in triplets_in_range(1, number // 2)
            if sum(triple) == number}


def triplets_in_range(start, end):
    for a in range(start, end):
        for b in range(a, end):
            for c in range(b, end):
                triple = (a, b, c)
                if is_triplet(triple):
                    yield triple


def is_triplet(triplet):
    a, b, c = triplet
    return (all(type(n) is int for n in triplet) and
            a ** 2 + b ** 2 == c ** 2)
